volumina/pixelpipeline/datasources/cachesource.py

```diff
@@ -0,0 +1,123 @@
+import itertools
+import threading
+import sys
+
+from PyQt5.QtCore import QObject, pyqtSignal
+
+from volumina.pixelpipeline.interface import DataSourceABC
+from volumina.slicingtools import is_pure_slicing
+from volumina.utility.cache import KVCache
+from volumina.config import CONFIG
+
+
+ARRAY_CACHE = KVCache(CONFIG.cache_size, getsizeof=sys.getsizeof)
+
+
+class _Request:
+    def __init__(self, cached_source, slicing, key):
+        self._cached_source = cached_source
+        self._slicing = slicing
+        self._key = key
+        self._result = None
+        self._rq = self._cached_source._source.request(self._slicing)
+
+    def wait(self):
+        if self._result is None:
+            self._result = res = self._rq.wait()
+            self._cached_source._cache[self._key] = res
+            self._cached_source._req.pop(self._key, None)
+
+        return self._result
+
+    def cancel(self):
+        self._rq.cancel()
+
+
+class _CachedRequest:
+    def __init__(self, result):
+        self._result = result
+
+    def wait(self):
+        return self._result
+
+    def cancel(self):
+        pass
+
+
+class CacheSource(QObject, DataSourceABC):
+    isDirty = pyqtSignal(object)
+    numberOfChannelsChanged = pyqtSignal(int)
+
+    __counter = itertools.count()
+
+    def __init__(self, source):
+        super().__init__()
+        self._lock = threading.Lock()
+        self._id = next(self.__counter)
+
+        self._source = source
+        self._cache = ARRAY_CACHE
+        self._req = {}
+        self._source.isDirty.connect(self.isDirty)
+        self._source.numberOfChannelsChanged.connect(self.numberOfChannelsChanged)
+        self._source.isDirty.connect(self.clear)
+        self._source.numberOfChannelsChanged.connect(self.clear)
+
+    def clear(self, *args):
+        self._cache.clear()
+        self._req.clear()
+
+    def __cache_key(self, slicing):
```

Comment:
We could use `(self._id, *slicing)` as a cache key, maybe with an additional check that `all(isinstance(s, collections.Hashable) for s in slicing)`. We also could use `id(self)` instead of `self._id = next(self.__counter)`.

Reply:
I don't see a reason to allow non string types as cache keys, in this case.